bridgepoint/prebuild.py

```diff
@@ -1005,11 +1013,21 @@ def accept_BinaryOperationNode(self, node):
         v_val_l = self.accept(node.left)
         v_val_r = self.accept(node.right)
 
+        l_s_dt = one(v_val_l).S_DT[820]()
         if node.operator.lower() in ['<', '<=', '==', '!=', '>=', '>',
                                      'and', 'or']:
             s_dt = self.s_dt('boolean')
+        elif node.operator.lower() in ['|', '+', '&', '^', '-'] and (
+            l_s_dt == self.s_dt('inst_ref<Object>') or l_s_dt == self.s_dt('inst_ref_set<Object>')):
```

Comment:
superfluous statement?

Reply:
meaning, l_s_dt is not used after this reassignment

Reply:
No reassignment here. l_s_dt is assigned on line 1016 and referenced here and line 1030, but never reassigned.

Reply:
Hmmm, actually there may be a problem here. It does not consider the case where a set operation is performed against two single instances of a specific class. For example:
```
select any inst1 from instances of OBJECT;
select any inst2 from instances of OBJECT;
inst_set = inst1 | inst2;
```
If "references are published", then the l_val type will be `inst_ref<OBJECT>` and the r_val type will also be `inst_ref<OBJECT>`. The binary operation type should be `inst_ref_set<OBJECT>`, however the else block will be taken and the type will be `inst_ref<OBJECT>`

Reply:
yea, right! 

It is a bit awkward to encode type information in a string. I believe this needs a bit more attention, possibly (as several incremental) improvements to the ooaofooa class model to facilitate better type inference.

Also (as TODO comments in pyxtuml suggests), coersion could improve the situation. Especially now when range types are in place in ooaofooa. See e.g. my old pascal frontend for llvm on range types: https://github.com/john-tornblom/llvm-p86/blob/master/llvm_p86/typesys.py#L113